File: apps/nest/deploy/infrastructure/snapshot_repo.go
package infrastructure

import (
	"encoding/json"
	"fmt"

	"github.com/koyeo/nest/deploy/domain"
)

const (
	nestMetaDir      = ".nest"
	snapshotFileName = ".nest/snapshot.json"
)

// SnapshotRepo implements domain.SnapshotRepository using RemoteFS.
type SnapshotRepo struct {
	fs domain.RemoteFS
}

// NewSnapshotRepo creates a new SnapshotRepo.
func NewSnapshotRepo(fs domain.RemoteFS) *SnapshotRepo {
	return &SnapshotRepo{fs: fs}
}

func (r *SnapshotRepo) Read(targetDir string) (*domain.Snapshot, error) {
	snapshotPath := fmt.Sprintf("%s/%s", targetDir, snapshotFileName)

	// Check if file exists
	_, err := r.fs.Stat(snapshotPath)
	if err != nil {
		// File does not exist
		return nil, nil
	}

	data, err := r.fs.ReadFile(snapshotPath)
	if err != nil {
		return nil, fmt.Errorf("read snapshot error: %s", err)
	}

	var snapshot domain.Snapshot
	if err = json.Unmarshal(data, &snapshot); err != nil {
		return nil, fmt.Errorf("decode snapshot error: %s", err)
	}
	return &snapshot, nil
}

func (r *SnapshotRepo) Write(targetDir string, snapshot *domain.Snapshot) error {
	metaDir := fmt.Sprintf("%s/%s", targetDir, nestMetaDir)

	// Ensure .nest directory exists
	if err := r.fs.MkdirAll(metaDir); err != nil {
		return fmt.Errorf("create .nest dir error: %s", err)
	}

	data, err := json.MarshalIndent(snapshot, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal snapshot error: %s", err)
	}

	snapshotPath := fmt.Sprintf("%s/%s", targetDir, snapshotFileName)
	if err = r.fs.WriteFile(snapshotPath, data); err != nil {
		return fmt.Errorf("write snapshot error: %s", err)
	}
	return nil
}

File: apps/ship/src/deploy/infrastructure/snapshot-repo.ts
import { REMOTE_META_DIR, REMOTE_SNAPSHOT_FILE } from "../../common/const.js";
import type { RemoteFS, SnapshotRepository } from "../domain/interfaces.js";
import { type Snapshot, snapshotSchema } from "../domain/snapshot.js";

export class SnapshotRepo implements SnapshotRepository {
  constructor(private readonly fs: RemoteFS) {}

  async read(targetDir: string): Promise<Snapshot | null> {
    const path = `${targetDir}/${REMOTE_SNAPSHOT_FILE}`;
    if ((await this.fs.stat(path)) === null) {
      return null;
    }
    const text = await this.fs.readFile(path);
    let parsed;
    try {
      parsed = JSON.parse(text);
    } catch (e) {
      throw new Error(`decode snapshot error: ${e instanceof Error ? e.message : String(e)}`);
    }
    const result = snapshotSchema.safeParse(parsed);
    if (!result.success) {
      throw new Error(`decode snapshot error: ${result.error.message}`);
    }
    return result.data;
  }

  async write(targetDir: string, snapshot: Snapshot): Promise<void> {
    await this.fs.mkdirAll(`${targetDir}/${REMOTE_META_DIR}`);
    await this.fs.writeFile(`${targetDir}/${REMOTE_SNAPSHOT_FILE}`, JSON.stringify(snapshot, null, 2));
  }
}
